### PR TITLE
Fixing System.Net.Security test failures in OS X

### PR DESCRIPTION
**Pull request by shrutigarg**
OS X has openssl 0.9.8 where TLSv1_1_method and TLSv1_2_method are not found so separating out the tls11 and tls12 related tests.

### CODE REVIEW

**Comment by shrutigarg** [comment]
cc: @vijaykota  @stephentoub  @CIPop

**Comment by shrutigarg** [comment]
fix for issue #4301

File: src/System.Net.Security/tests/FunctionalTests/ClientAsyncAuthenticateTest.cs [hunk 1]
```diff
@@ -72,6 +69,20 @@ private static IEnumerable<object[]> ProtocolMismatchDataSsl2SpecificLinux()
...
+        }
+
+        private static IEnumerable<object[]> ProtocolMismatchData_Tls11_Tls12_OSX()
+        {
+            yield return new object[] { SslProtocols.Tls11, SslProtocols.Ssl2, typeof(AuthenticationException) };
+            yield return new object[] { SslProtocols.Tls11, SslProtocols.Tls12, typeof(AuthenticationException) };
+            yield return new object[] { SslProtocols.Tls12, SslProtocols.Ssl2, typeof(AuthenticationException) };
+        }
+
         public ClientAsyncAuthenticateTest()
```
Comment (eerhardt) [inline src/System.Net.Security/tests/FunctionalTests/ClientAsyncAuthenticateTest.cs]: Should this method have a [Fact] attribute on it?

**Comment by eerhardt** [comment]
LGTM.
Does this completely fix #4301?

**Comment by CIPop** [comment]
Have you considered an exception adapter that would bring all platforms to the same behavior (at least when SslStream is used in server mode)? What are the advantages/disadvantages of making the exceptions uniform?

Otherwise, LGTM

**Comment by stephentoub** [comment]
> Have you considered an exception adapter that would bring all platforms to the same behavior (at least when SslStream is used in server mode)?

We should bring them all in line.  I asked about this in a previous PR, and I didn't really understand the answer about why there needed to be differences.  Unless it's really problematic to have the same exception types being used (in which case I'd like to understand that), we should be using the same exception types for the same situations across platforms.

**Comment by vijaykota** [comment]
@shrutigarg, are we getting different exceptions in the SSL2 mismatch test cases because on Windows, the handshake fails while on Linux/OSX, the client-side fails before the handshake due to missing SSL 2 (and TLS 1.1/2)  support?

**Comment by vijaykota** [comment]
> Does this completely fix #4301?

@eerhardt, 4301 also includes a failure seen on Linux - this may be same as #4317 and is yet to be investigated

**Comment by shrutigarg** [comment]
@stephentoub @CIPop 
 For any failure scenario System.Net.Security code throws 2 types of exception IOException and AuthenticationException.
IOException is thrown when in server-client communication , processing of handshake message false  at server end and server terminates. Due to which client does not get any response backand result  IOException.
Whereas Authentication exception is thrown when client itself fails at any stage being the server active.
So in the scenarios where we are getting different exceptions on different platforms are:

Client is SSL2 :
For client being ssl2 ,Windows and Unix code behave differently. In case of Windows (where SSL2 is supported) when client sends "SSL2 Client Hello", Server rejects this client hello message as it is enabled with any other protocol. As a result it terminates itself and at client side we get IO Exception.
Whereas in case of Unix (Where ssl2 is not supported and we enable SSl23_Method() in our getMethod implementation for ssl2 protocol flag) client sends the "SSL23 Client Hello" which actually (we got to know by looking at the packet) sends SSL3 packet bytes since creating ssl2 hello message was not supported .Seems a fallback logic in OpenSsl .
As a result server accepts this Client Hello and responds back with its Server Hello with the protocol server is enabled with. And now client side fails in parsing the "server Hello"  with unsupported Protocol by comparing the ssl protocol versions server/client were enabled with . So we get Authentication exception. 
SO if we want to match the exception type , we would have to expect IO Exception in case of Unix which is not possible as server doesn't get terminated. (something which is by design in Openssl implementation if we enable SSL23_method when ssl2 protocol option is set).
What we can change at our end is, not enabling ssl23_Method but return a nullptr as method for ssl2 protocol flag (the way we are doing for TLS1.1 TLS1.2 for OS X) In that case client will fail before even sending the hello packet of any other protocol). But that case too we will get the Authentication Exception.

Client is TLS1.1/ TLS1.2 :
Similarly it is happening in case of OSX. Since client side itself fails when null method is returned in case of TLS1.*, We expect the Authentication Exception ..

In short, the failure is happening at different ends (in windows server side and in Unix/OSX client side), we are expecting different type of exceptions.

Still if we see a major issue in this , the only way to force the exception to be same by putting some hack  in CheckCompletionBeforeNextReceive() to generate the IOException by looking at some type of error codes which message.GetException() will return . Which would not be correct.

**Comment by CIPop** [comment]
@shrutigarg Thanks for the in-detail explanation: it makes sense to differentiate between server-side responses. 
You are correct in not adapting the different exceptions if the underlying OS implementation's behavior cannot be changed.

LGTM
